db/migrations/20200305202405_create_get_or_create_header_function.sql

```diff
@@ -0,0 +1,43 @@
+-- +goose Up
+-- SQL in this section is executed when the migration is applied.
+
+-- +goose StatementBegin
+CREATE OR REPLACE FUNCTION public.get_or_create_header(block_number BIGINT, hash VARCHAR(66), raw JSONB, block_timestamp NUMERIC, eth_node_id INTEGER) RETURNS INTEGER AS
+$$
+DECLARE
+    matching_header_id INTEGER := (
+        SELECT id
+        FROM public.headers
+        WHERE headers.block_number = get_or_create_header.block_number
+        AND headers.hash = get_or_create_header.hash
+        );
+    nonmatching_header_id INTEGER := (
+        SELECT id
+        FROM public.headers
+        WHERE headers.block_number = get_or_create_header.block_number
+        AND headers.hash != get_or_create_header.hash
+        );
+    inserted_header_id INTEGER;
+BEGIN
+    IF matching_header_id != 0 THEN
+        RETURN matching_header_id;
+    END IF;
+
+    IF nonmatching_header_id != 0 THEN
+        DELETE FROM public.headers WHERE id = nonmatching_header_id;
```

Comment:
One thing I want to work on making super clear is that we need to delete, rather than update, a header with the wrong hash so that we trigger the cascade delete behaviors on events, etc.